File: array/insert-interval.js
// 57. Insert Interval
// https://leetcode.com/problems/insert-interval/description/

/**
 * @param {number[][]} intervals
 * @param {number[]} newInterval
 * @return {number[][]}
 */
var insert = function (intervals, newI) {
  const stack = [];

  for (const currI of intervals) {
    if (newI[0] > currI[1]) stack.push(currI);
    else if (newI[1] < currI[0]) {
      stack.push(newI);
      newI = currI;
    } else newI = [Math.min(newI[0], currI[0]), Math.max(newI[1], currI[1])];
  }

  stack.push(newI);

  return stack;
};

// Time complexity: O(n) -> We iterate over the array of intervals once.
// Space complexity: O(n) -> The stack holds at most n + 1 intervals, where n is the number of original intervals.

// Initialize an empty stack to hold the resulting intervals.
// Iterate through the intervals one by one:
// - If newInterval starts after the current interval ends (no overlap), push the current interval to the stack.
// - If newInterval ends before the current interval starts (no overlap), push newInterval first, then update newInterval to be the current interval.
// - If there's overlap between newInterval and the current interval, merge them by taking the minimum start time and maximum end time.
// After the loop finishes, the last interval (newInterval) will need to be pushed to the stack since it hasn't been added yet.

File: insert-interval.java
// https://leetcode.com/problems/insert-interval/description/

// Java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        ArrayList<int[]> list = new ArrayList<>();
        int[] toAdd = newInterval;

        for (int[] i : intervals){
            // Case 1: i comes before newInterval
            if (toAdd[0] > i[1]) list.add(i);

            // Case 2: i comes after newInterval{}
            else if (toAdd[1] < i[0]) {
                list.add(toAdd);
                toAdd = i;
            }

            // Case 3: overlapping interval
            else {
                toAdd[0] = Math.min(toAdd[0], i[0]);
                toAdd[1] = Math.max(toAdd[1], i[1]);
            }
        }

        list.add(toAdd);

        return list.toArray(new int[list.size()][]);
    }
}

// JS
var insert = function(intervals, newInterval) {
    let toAdd = newInterval
    let res = []

    for (let i of intervals){
        // case 1: i before toAdd
        if (toAdd[0] > i[1]) res.push(i);

        // case 2: i after toAdd
        else if (toAdd[1] < i[0]){
            res.push(toAdd)
            toAdd = i
        }

        // case 3: overlapping
        else {
            toAdd[0] = Math.min(toAdd[0], i[0])
            toAdd[1] = Math.max(toAdd[1], i[1])
        }
    }

    res.push(toAdd)

    return res
};

// for any interval in the input, it can
    // come before
    // come after
    // overlap
// create a placehholder [][] to hold value -> init to newInterval
// iterate through intervals i
    // if i happens before placeholder, simply add i to the res
    // if i happens after placeholder, add placeholder to res + updated placeholder to i
    // if overlap, udate the placeholder